keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,1699 @@
++       // QoSResources specifies the QoS resources.
++       // +optional
++       QoSResources map[QoSResourceName]string
+ }
+
++// QoSResourceName is the name of a QoS resource.
++type QoSResourceName string
+```
+
+Also, we add a Resources field to the PodSpec, to enable assignment of
+pod-level QoS resources. We will re-use the existing ResourceRequirements type
+but Limits and Requests and Claims must be left empty. QoSResources may be set
+and they represent the Pod-level assignment of QoS-class resources,
+corresponding the PodQoSResources message in PodSandboxConfig in the CRI
+API.
+
+```diff
+ type PodSpec struct {
+@@ -3062,6 +3069,10 @@ type PodSpec struct {
+        ResourceClaims []PodResourceClaim
++       // Pod-level resources. Claims, requests and limits are not allowed
++       // to be specified for pods.
++       // +optional
++       Resources ResourceRequirements
+ }
+```
+
+There is already an ongoing effort to add [Pod level resource limits][kep-2837]
+that aims at adding a pod level `Resources` field in a similar fashion. Thus,
+we opt for adding ResourceRequirements insted of QoSResources directly into the
+PodSpec.
+
+As an example, a Pod requesting class "fast" of a (exemplary) pod-level QoS
+resource named "network", with one container requesting class "gold" of
+container-level QoS resource named "rdt":
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: qos-resource-example
+spec:
+  resources:
+    qosResources:
+      network: fast
+  containers:
+  - name: cnt
+    image: nginx
+    resources:
+      qosResources:
+        rdt: gold
+```
+
+#### NodeStatus
+
+We extend NodeStatus to list available QoS-class resources on a node, This
+consists of the list of the available QoS-class resource types and the classes
+available within each of these resource types.
+
+
+
+```diff
+ type NodeStatus struct {
+@@ -4444,6 +4482,11 @@ type NodeStatus struct {
+        // Status of the config assigned to the node via the dynamic Kubelet config feature.
+        Config *NodeConfigStatus
++       // QoSResources contains information about the QoS resources that are
++       // available on the node.
++       // +featureGate=QoSResources
++       // +optional
++       QoSResources QoSResourceStatus
+ }
+
+...
+
++// QoSResourceStatus describes QoS resources available on the node.
++type QoSResourceStatus struct {
++       // PodQoSResources contains the QoS resources that are available for pods
++       // to be assigned to.
++       PodQoSResources []QoSResourceInfo
++       // ContainerQoSResources contains the QoS resources that are available for
++       // containers to be assigned to.
++       ContainerQoSResources []QoSResourceInfo
++}
+
++// QoSResourceInfo contains information about one QoS resource type.
++type QoSResourceInfo struct {
++       // Name of the resource.
++       Name QoSResourceName
++       // Mutable is set to true if the resource supports in-place updates.
++       Mutable bool
```

Comment:
What happens if a class is removed while pods are using it?

Reply:
I [replied](https://github.com/kubernetes/enhancements/pull/3004#issuecomment-1414043336) to @SergeyKanzhelev regarding pod evicition. I think we should treat QoS resources similarly to extended resources (but I'm now unsure how that behaves). For the current usage scenarios pod eviction hasn't been "interesting" but maybe it will be for others. -> will add this to the future work section

Reply:
ping @thockin ^